### PR TITLE
Explicitly allow IPIP packets from/to Calico hosts.

### PR DESCRIPTION
**Pull request by fasaxc**
## Description
Rather than fixing https://github.com/projectcalico/calico/issues/888 by doccing that a rule is needed, maybe we should make felix auto-generate the rule?  WDYT?

This removes the need for an explicit policy rule when using IPIP and host protection together.

Fixes https://github.com/projectcalico/calico/issues/888
Fixes https://github.com/projectcalico/calico/issues/1357

## Todos
- [x] Unit tests (full coverage)
- [x] Integration tests (delete as appropriate) In plan/Not needed/Done
- [x] Documentation - checked the host protection guide, doesn't mention IPIP
- [x] Release note

## Release Note

```release-note
When IPIP is enabled, Calico now installs an explicit allow rule for its IPIP traffic.  This removes the need to manually add such a rule when using host protection policy. 
```

### CODE REVIEW

**Comment by Dirbaio** [comment]
+1000 to this. Our prod cluster had IPIP + host endpoint policy and I overlooked the "allow IPIP" detail. The resulting issues were very hard to debug, since conntrack would allow incoming IPIP packets if the host had sent an outgoing IPIP packet before. It would only fail when spinning up a new node, only before that node had tried to send out IPIP traffic.

re: the security concern: IIRC IPIP traffic goes through iptables two times, first wrapped, then unwrapped, so you can't circunvent hostendpoint policies sending traffic through the tunnel destined for the node's IP.

**Comment by fasaxc** [comment]
Yes, I think @Dirbaio is correct, the unwrapped packet is re-injected and seems to come from the tunl0 device, where our other policy will get applied as normal.

File: fv/containers/containers.go [hunk 1]
```diff
@@ -308,3 +312,84 @@ func StartSingleNodeEtcdTopology() (felix, etcd *Container, client client.Interf
 	success = true
 	return
 }
+
+// StartSingleNodeEtcdTopology starts an etcd container and a single Felix container; it initialises
```
Comment (neiljerram) [inline fv/containers/containers.go]: Single -> Two

Comment (neiljerram) [reply fv/containers/containers.go]: (also "a single Felix container")

File: fv/containers/containers.go [hunk 2]
```diff
@@ -308,3 +312,84 @@ func StartSingleNodeEtcdTopology() (felix, etcd *Container, client client.Interf
...
+// the datastore and installs a Node resource for the Felix node.
+func StartTwoNodeEtcdIPIPTopology() (felixes []*Container, etcd *Container, client client.Interface) {
+	log.Info("Starting a single-node etcd topology.")
```
Comment (neiljerram) [inline fv/containers/containers.go]: single again

File: fv/containers/containers.go [hunk 3]
```diff
@@ -308,3 +312,84 @@ func StartSingleNodeEtcdTopology() (felix, etcd *Container, client client.Interf
...
+		}
+		ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		ipPool := api.NewIPPool()
+		ipPool.Name = "test-pool"
+		ipPool.Spec.CIDR = "10.65.0.0/16"
+		ipPool.Spec.IPIPMode = api.IPIPModeAlways
+		_, err = client.IPPools().Create(ctx, ipPool, options.SetOptions{})
+		return err
+	}).ShouldNot(HaveOccurred())
```
Comment (neiljerram) [inline fv/containers/containers.go]: I guess all that is shared with the single node setup case - can it be commonized?

Comment (neiljerram) [reply fv/containers/containers.go]: Or you could just make the number of felixes a parameter to this whole function, and use it in the single case too.

File: fv/containers/containers.go [hunk 4]
```diff
@@ -308,3 +312,84 @@ func StartSingleNodeEtcdTopology() (felix, etcd *Container, client client.Interf
...
+		felixes = append(felixes, felix)
+	}
+
+	for i, iFelix := range felixes {
+		for j, jFelix := range felixes {
+			if i == j {
+				continue
+			}
+
+			jBlock := fmt.Sprintf("10.65.%d.0/24", j)
```
Comment (neiljerram) [inline fv/containers/containers.go]: Possible comment about not using IPAM here.

File: fv/ipip_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+				_, err := client.HostEndpoints().Create(ctx, hep, options.SetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+			}
+		})
+
+		It("should have workload connectivity but not host connectivity", func() {
+			// Host endpoints block host-host traffic.
+			cc.ExpectNone(felixes[0], hostW[1])
+			cc.ExpectNone(felixes[1], hostW[0])
+			// But the rules to allow IPIP between our hosts let the workload traffic through.
```
Comment (neiljerram) [inline fv/ipip_test.go]: Add "and the lack of applyOnForward"?

File: felix.go [hunk 6]
```diff
@@ -639,18 +653,36 @@ func monitorAndManageShutdown(failureReportChan <-chan string, driverCmd *exec.C
...
 	}
 
-	if !receivedSignal {
+	if !receivedFatalSignal {
 		// We're exiting due to a failure or a config change, wait
 		// a couple of seconds to ensure that we don't go into a tight
-		// restart loop (which would make the init daemon give up trying
-		// to restart us).
-		logCxt.Info("Shutdown wasn't caused by signal, pausing to avoid tight restart loop")
+		// restart loop (which would make the init daemon in calico/node give
```
[the remark this answers is not in the document]
Comment (fasaxc) [reply felix.go]: I'm talking about the init daemon used in calico/node, which is `runit`, not `tini`, and it does restart `calico-felix`

[the remark this answers is not in the document]
Comment (neiljerram) [reply felix.go]: Ah, right, sorry.

File: fv/ipip_test.go (hunk 5 above)
Comment (fasaxc) [reply fv/ipip_test.go]: Discussed offline and improved surrounding comment.

**Pull request merged**